calblueprint/meepanyar
# Add CustomerData slice to redux

[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
This PR is a refactor of our redux store to add a `CustomerDataSlice`. This was done for the following reasons
1. To make the division of data in our redux store nicer (less need to nest deeply into `siteData` to grab properties)
2. Since Redux data is immutable, un-nesting our data is advantageous because we do less copying when we persist new data (such as a customer or meter reading) into the redux store
3. There was some trouble with the data flow in getting the customer in `CustomerProfile` to the `Edit Customer` screen. We remedy this by adding a global `currentCustomerId` to the redux store, which will be the customer that will be edited or displayed when navigating to the `Edit Customer` or `CustomerProfile` screens respectively.
    - **Note**: We should standardize the ^ usage if we adopt it. That is, we should no longer pass state via Link for things like Customer, and we should be consistent with always grabbing the `currentCustomer` via the redux store.

TODO BEFORE MERGE: Recreate patch for request.js

## Changelog
- Create a `customerData` slice in Redux with a `currentCustomer` attribute
- Created getter and setter methods for customerData
- `/customers/customer` now displays whatever the customer corresponding to`customerData.currentCustomerId` from redux. `Link` and `addCustomer` functionality was changed to set this value before navigating to `CustomerProfile`
- No functional changes should be observed (only state management)
- Persist IDs from Airtable for Customers who've just been created
- Airtable customers made while the app was offline will receive a temporary offline ID
- Changed `EditCustomer` to `goBack` instead of replace

**The backend branch `julian/create-customer-id` MUST BE USED for this PR**
PR for backend: https://github.com/calblueprint/meepanyar-node/pull/18

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

## How to review
- Make sure you've checked out the correct backend branch from the PR [here](https://github.com/calblueprint/meepanyar-node/pull/18)
- Log in and check the redux store's `customerData` is populating correctly.
- Navigate to all customer-related links and make sure they work properly
- Create a new customer and ensure sure that flow still works properly
  - You are taken to the new customer's profile
  - Going back to the customer list, you can find the customer that was just created
  - Attempt to "Edit" the customer you just created. After editing, ensure that the edits are accurately reflected on `customerMain`, `customerProfile` and on the airtable base.

[//]: # 'The order in which to review files and what to expect when testing locally'

## Relevant Links

### Online sources
https://redux-toolkit.js.org/api/createEntityAdapter
https://redux-toolkit.js.org/api/createSelector

[//]: # 'Optional - copy links to any tutorial or documentation that was useful to you when working on this PR'

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## Next steps
- Refactor `siteData` and add useful `getter` functions for that
- Make offline ID generate robust
- Persist IDs from Airtable for Customers who've just been created
- Debatable: Refactor `customerData` to a `Record<SiteId, Record<CustomerId, CustomerRecord>>` once offline IDs are added
- Use Selectors for `getter` functions like `getCurrentCustomer` and `getAllCustomersInSite`
- Potentially switch to [createEntityAdapter](https://redux-toolkit.js.org/api/createEntityAdapter) in the future

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

### Screenshots

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @julianrkung

[//]: # 'This tags Julian as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
